packages/wallet/contract/src/smartWallet.js

```diff
@@ -1,90 +1,273 @@
+    });
+    pursePublishKit.publisher.publish(record);
+  };
+
+  /** @type {PublishKit<OffersRecord>} */
+  const offerPublishKit = makePublishKit();
+  const offersSubscriber = makeStoredSubscriber(
+    offerPublishKit.subscriber,
+    E(myWalletStorageNode).makeChildNode('offers'),
     marshaller,
   );
+  /**
+   * @param {import('./offers.js').OfferStatus} status
+   */
+  const publishOfferStatus = status => {
+    const record = { [status.id]: status };
+    offerPublishKit.publisher.publish(record);
+  };
+  // #endregion
+
+  // #region issuer management
+  /** @type { (desc: BrandDescriptor) => Promise<void>} */
+  const addIssuer = async desc => {
+    // assert haven't received this issuer before.
+    const descriptorsHas = brandDescriptors.has(desc.brand);
+    const pursesHas = brandPurses.has(desc.brand);
+    assert(
+      !(descriptorsHas && pursesHas),
+      'repeated brand from bank asset subscription',
+    );
+    assert(
+      !(descriptorsHas || pursesHas),
+      'corrupted state; one store has brand already',
+    );
+
+    // make a purse for this issuer.
+    // first make the purse so that if anything fails we don't update stores
+    const purse = await E(desc.issuer).makeEmptyPurse();
+
+    // save all five of these in a collection (indexed by brand?) so that when
+    // it's time to take an offer description you know where to get the
+    // relevant purse. when it's time to make an offer, you know how to make
+    // payments. REMEMBER when doing that, need to handle every exception to
+    // put the money back in the purse if anything fails.
+    brandDescriptors.init(desc.brand, desc);
+    brandPurses.init(desc.brand, purse);
+
+    publishPurses();
+  };
+
+  // Ensure a purse for each issuer
+  addIssuer({
+    brand: invitationBrand,
+    issuer: invitationIssuer,
+    petname: 'invitations',
+  });
+  // watch the bank for new issuers to make purses out of
+  observeIteration(E(bank).getAssetSubscription(), {
+    async updateState(desc) {
+      await addIssuer({
+        brand: desc.brand,
+        issuer: desc.issuer,
+        petname: desc.proposedName,
+      });
+    },
+  });
+  // #endregion
+
+  /**
+   * Similar to {DepositFacet} but async because it has to look up the purse.
+   */
+  const depositFacet = Far('smart wallet deposit facet', {
+    /**
+     * Put the assets from the payment into the appropriate purse
+     *
+     * @param {ERef<Payment>} paymentE
+     * @param {Brand} paymentBrand must match the payment's brand. passed in to save lookup.
+     * @returns {Promise<Amount>}
+     * @throws if the purse doesn't exist
+     * ??? PRODUCT: should it instead wait for an issuer/purse to put it in?
+     */
+    receive: (paymentE, paymentBrand) => {
+      const purse = brandPurses.get(paymentBrand);
+
+      const depositP = E.when(paymentE, payment => E(purse).deposit(payment));
+
+      E.when(depositP, publishPurses, () =>
```

Comment:
the purse's balance notifier should suffice; I don't think we need to `publishPurses` here.

Reply:
Great. I didn't know about that notifier but it helps a lot. I'll refactor using it after I get the bridge working.